### ISSUE TITLE
ArkBot performance issues

### ISSUE THREAD

**Issue by smooreace** [ark-mod/ArkBot#75]
The best way to describe this is... CPU hungry beast mode!

After running for a few days I start having severe server performance issues. ArkBot pounds the living hell out of the CPU, but memory and disk activity are essentially unchanged. The images taken are from a period of time where no one has direct access to anything ArkBot provides except Cross Ark Chat. Ports are disabled to prevent any access to the website side of things beyond me. And I haven't opened it since ArkBot was started. Config file provided via pastebin.

https://pastebin.com/FshT9wqY

https://gyazo.com/49976e3f70abe3ae22513f5d7586d7d3

https://gyazo.com/5af5fb1a0f457bc42f84a122d71758c6

**Comment by jefferson-1** [comment]
We are aware of this issue. It is on the roadmap to add the ability to affine the process to particular CPU threads and not to all threads.

Affinity can be set manually... https://www.tech-recipes.com/rx/37272/set-a-programs-affinity-in-windows-7-for-better-performance/

**Comment by Magesw** [comment]
Use process lasso, will take care of that real fast. I run 6 ARK servers and a couple other things with ARKBot on the same server, process lasso takes care of it all.

**Comment by smooreace** [comment]
Yes... I can tone it down to only have access to maybe one core. But it already performs INSANELY slow on the website side of things. I'm running Arkbot on 2 servers, each with 12 or so Ark servers. Maybe I'm pushing Arkbot further than its intended? That's 2 clusters per instance.

**Comment by jefferson-1** [comment]
The code for ARK is already cumbersome. Adding ArkBot further complicates things. ARKBot when it runs does a lot of things that is processor intensive. 

To resolve it on our cluster, I threw a ThreadRipper with 64gb of RAM at it. That solved it for our 8 cluster at knightsgg.com. I was able to lock it down to 4 cores and it seems happy.

**Comment by Magesw** [comment]
For Dino Fight Club I just lock it to one core and it does fine with my one cluster and my other box running the other cluster I do the same, one core.

**Comment by smooreace** [comment]
Im testing it now with 2 cores. Its behaving ok at the moment... but time will tell. I wish the damn thing wasnt so memory hungry over time!! bEtween the memory leak and CPU insanity its more work to manage than the 24 Ark servers!

**Comment by smooreace** [comment]
Thanks for the info everyone

**Comment by Magesw** [comment]
I have not seen a memory leak issue yet though with mine.

**Comment by smooreace** [comment]
Now, after running for about 5 minutes, all server have been updated. It theoretically should be done with its initial work...

Both instances are running pretty stable at just under 1GB of RAM usage. By day 5 or so its up to almost 4GB. Ive seen it as high as 10GB. But I dont let it run that long any more.

**Comment by smooreace** [comment]
I lied... both instances were started at the same time. Both managing about the same amount of servers. But both are vastly different in memory usage. One has shot up quite quickly

https://gyazo.com/6856731c5c6905206e9c0c77fc5cf709

https://gyazo.com/1f3929c04055d716d7d6aca6054fb16e

**Comment by smooreace** [comment]
The one using less RAM is managing MORE servers too!

**Comment by Magesw** [comment]
Weird I wonder if it the version of windows you are using because my ARKBot has been running for at least a week straight and I just looked at it's memory usage 
https://prnt.sc/kwbas6

**Comment by Magesw** [comment]
I am using Windows server 2016.

**Comment by smooreace** [comment]
Wow... ya! Thats a lot better. But I am pushing mine a lot harder. Running under Server 2008. Not that it matters, but each machine has 11 to 15 Ark servers active, plus Conan, DnL, and PixArk servers running. Its not like memory is particularly an issue. But I dont like things to get out of hand. The CPU issue was the real killer.

**Comment by Magesw** [comment]
That is with a threadripper CPU? I used to run mine on a 1700x I loved that CPU, now I am on OCed i7's.

**Comment by smooreace** [comment]
Dual X5570

**Comment by tsebring** [comment]
To add some context to the performance issues of ArkBot; most of the CPU-cycles are spent extracting savegame data and rebuilding the .NET domain model for that same data. It is an inherently costly and time-consuming process.

When the application was originally built it used the Java version of [ark-tools](https://github.com/Qowyn/ark-tools) to extract savegame data to JSON. This was, for reasons I can't recall, too slow to work well with multiple servers so I translated [ark-savegame-toolkit](https://github.com/Qowyn/ark-savegame-toolkit) into .NET where I could more easily make targeted optimizations - [ArkSavegameToolkitNet](https://github.com/ark-mod/ArkSavegameToolkitNet).

Qowyn, the author of ark-tools, have since then made massive improvements to the codebase and the extraction speed have far surpassed my .NET implementation including optimizations. Due to limited time I was never able to bring those improvements over to [ArkSavegameToolkitNet](https://github.com/ark-mod/ArkSavegameToolkitNet).

Around the same time, the road forward that I wanted to take changed from extracting the savegame data to fetching it straight from memory using custom plugins for [ARK Server API](https://arkserverapi.com/).

That is where we are today - a lot of the CPU-cycles are wasted due to [ArkSavegameToolkitNet](https://github.com/ark-mod/ArkSavegameToolkitNet) being built on an old version of [ark-tools](https://github.com/Qowyn/ark-tools). There are a lot of possibilities for improvements however the work involved should not be underestimated.

**Comment by smooreace** [comment]
Thanks for the info! Please don't take my issue report as a complaint! I would FAR rather it run poorly and work, than to not exist at all!!

**Comment by ark-mod** [comment]
> Thanks for the info! Please don't take my issue report as a complaint! I would FAR rather it run poorly and work, than to not exist at all!!

No, not at all! 😃 It is so that other devs know the history and current state of the application.

**Comment by tsebring** [comment]
> Thanks for the info! Please don't take my issue report as a complaint! I would FAR rather it run poorly and work, than to not exist at all!!

No, not at all! 😃 It is so that other devs know the history and current state of the application.

**Comment by smooreace** [comment]
As an update, Ive limited ArkBot to 2 cores per on each machine. Since it was started one is managing 11 servers and the other is managing 12. 

ArkBot on the 12 server machine is running 5.83 to 7.04% CPU and using 1.26 to 1.54 GB of RAM
ArkBot on the 11 server machine is running 0.02 to 10,63% CPU and a flat 1.27 GB RAM

Still can't make sense of the difference. The machines are identical except for the actual game servers being hosted. And the 11 server machine actually gets about 185% MORE play time than the 12 server machine. With that, there are FAR more structures and tames. Yet its using less resources.

**Comment by tsebring** [comment]
CPU% does not really tell you that much - could just be other threads using those same cores or lots of switching happening. You'd be better off comparing CPU time.

Also check the physical size of the savegames and the number of profiles.

**Comment by smooreace** [comment]
Those stats are CPU usage for that process alone. If it were looking at 2 totally different machines I could see somewhat of an inaccuracy. But they are identical machines.
